src/common/transformations/src/transformations/op_conversions/einsum_decomposition.cpp

```diff
@@ -434,12 +642,273 @@ void reduce_input(ov::pass::EinsumDecomposition* einsum_decompose_ptr,
     subgraph_nodes.insert(subgraph_nodes.end(), {axes_const, reduce_sum});
 }
 
+/// \brief Prepares data for diagonal extraction in Einsum operation.
+///
+/// This function processes the input subscript and label-dimension map to identify repeated labels,
+/// update the resultant subscript, and determine the axes to be reduced.
+///
+/// \param input_subscript The input subscript string representing the Einsum equation.
+/// \param label_dim_map A map from labels to their corresponding dimensions.
+/// \param resultant_subscript A reference to the resultant subscript string to be updated.
+/// \param resultant_subscript_with_duplicates A reference to a resultant subscript string with duplicates.
+/// \param repeated_labels A reference to a vector of strings to store repeated labels found in input subscript.
+/// \param unrepeated_labels A reference to a vector where unrepeated labels will be stored.
+/// \param reduced_axes A reference to an AxisVector to store the axes that need to be reduced.
+///
+void prepare_diagonal_extraction_data(const std::string& input_subscript,
+                                      const LabelDimMap& label_dim_map,
+                                      std::string& resultant_subscript,
+                                      std::string& resultant_subscript_with_duplicates,
+                                      std::vector<std::string>& repeated_labels,
+                                      std::vector<std::string>& unrepeated_labels,
+                                      ov::AxisVector& reduced_axes) {
+    static const std::string ellipsis = "...";
+    const auto labels = ov::op::v7::Einsum::extract_labels(input_subscript);
+    std::vector<std::string> repeated_labels_with_duplicates;
+    size_t reduced_dim = 1;
+    for (const auto& label : labels) {
+        if (resultant_subscript.find(label) != std::string::npos) {
+            continue;
+        }
+
+        const auto dims_it = label_dim_map.find(label);
+        OPENVINO_ASSERT(dims_it != label_dim_map.end());
+
+        auto dims = dims_it->second;
+        const auto dims_size = dims.size();
+        OPENVINO_ASSERT(dims_size > 0);
+
+        if (label != ellipsis && dims_size > 1) {
+            // repeated label is found
+            // save only the first dimension corresponding to the repeated label
+            dims = {dims[0]};
+            reduced_axes.push_back(reduced_dim);
+            repeated_labels.push_back(label);
+            repeated_labels_with_duplicates.insert(repeated_labels_with_duplicates.end(), dims_size, label);
+            reduced_dim += 2;
```

Comment:
please double-check if it is `+2` required.

Reply:
It is, please check next comment for details. But in short, every second dimension should be reduced based on shape: 
`[a_diagonal, a_leftover_from_pad, b_diagonal, b_leftover_from_pad, ..., x_unrepeated_dim, y_unrepeated_dim]`
So {1, 3=2+1, 5=3+2,...} for each repeated label